Python/specialize.c

```diff
@@ -2071,7 +2071,7 @@ _Py_Specialize_BinaryOp(PyObject *lhs, PyObject *rhs, _Py_CODEUNIT *instr,
             if (PyUnicode_CheckExact(lhs)) {
                 _Py_CODEUNIT next = instr[INLINE_CACHE_ENTRIES_BINARY_OP + 1];
                 bool to_store = (next.op.code == STORE_FAST);
-                if (to_store && locals[next.op.arg] == lhs) {
+                if (to_store && PyStackRef_To_PyObject_Steal(locals[next.op.arg]) == lhs) {
```

Comment:
```suggestion
                if (to_store && UNDEFER(locals[next.op.arg]) == UNDEFER(lhs)) {
```
We shouldn't assume that `PyStackRef_To_PyObject_Steal(ref)` produces the same object every time. We want to compare references (without deferral).
`UNDEFER` would turn a stack ref into a heap ref, otherwise this will need to box any unboxed values.

`PyStackRef_To_PyObject_Steal` invalidates the reference it takes as an argument. You need to borrow the reference.